pkg/csi/service/wcpguest/controller.go

```diff
@@ -531,6 +575,75 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
+	timeoutSeconds := int64(getAttacherTimeoutInMin(ctx) * 60)
+	cnsFileAccessConfigInstanceName := req.NodeId + "-" + req.VolumeId
+	watchCnsFileAccessConfig, err := cnsFileAccessConfigWatcher.Watch(metav1.ListOptions{
+		FieldSelector:  fields.SelectorFromSet(fields.Set{"metadata.name": cnsFileAccessConfigInstanceName}).String(),
+		TimeoutSeconds: &timeoutSeconds,
+	})
+	if err != nil {
+		msg := fmt.Sprintf("failed to watch cnsFileAccessConfig instance %q/%q with Error: %v", c.supervisorNamespace, cnsFileAccessConfigInstanceName, err)
+		log.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	if err := c.cnsOperatorClient.Delete(ctx, &cnsfileaccessconfigv1alpha1.CnsFileAccessConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cnsFileAccessConfigInstanceName,
+			Namespace: c.supervisorNamespace,
+		},
+	}); err != nil {
+		msg := fmt.Sprintf("failed to delete CnsFileAccessConfig instance: %q/%q. Error: %+v", c.supervisorNamespace, cnsFileAccessConfigInstanceName, err)
+		log.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	defer watchCnsFileAccessConfig.Stop()
+	var cnsFileAccessConfigInstanceErr string
+	isCnsFileAccessConfigInstanceDeleted := false
+	// Watch all update events made on CnsFileAccessConfig instance until Deleted event or a timeout occurs on the cnsfileaccessconfig instance
+	for !isCnsFileAccessConfigInstanceDeleted {
+		log.Debugf("waiting for update on cnsfileaccessconfigs: %q", cnsFileAccessConfigInstanceName)
+		event := <-watchCnsFileAccessConfig.ResultChan()
+		cnsfileaccessconfig, ok := event.Object.(*cnsfileaccessconfigv1alpha1.CnsFileAccessConfig)
+		if !ok {
+			msg := fmt.Sprintf("Watch on cnsfileaccessconfig instance %q/%q timed out. Last seen error on the instance=%q", c.supervisorNamespace, cnsFileAccessConfigInstanceName, cnsFileAccessConfigInstanceErr)
+			log.Error(msg)
+			return nil, status.Errorf(codes.Internal, msg)
+		}
+		if cnsfileaccessconfig.Name != cnsFileAccessConfigInstanceName {
+			log.Debugf("Observed CnsFileAccessConfig instance name: %q, expecting CnsFileAccessConfig instance name: %q. Continuing...", cnsfileaccessconfig.Name, cnsFileAccessConfigInstanceName)
+			continue
+		}
+		// Check if SV PVC Name ain the cnsfileaccessconfig instance match with VolumeId from the request
+		if cnsfileaccessconfig.Spec.PvcName != req.VolumeId {
+			log.Debugf("Observed SV PVC Name: %q, expecting SV PVC Name: %q. Continuing...", cnsfileaccessconfig.Spec.PvcName, req.VolumeId)
+			continue
+		}
+		// Check if VM name in the cnsfileaccessconfig instance match with NodeId from the request
+		if cnsfileaccessconfig.Spec.VMName != req.NodeId {
+			log.Debugf("Observed vm name: %q, expecting vm name: %q. Continuing...", cnsfileaccessconfig.Spec.VMName, req.NodeId)
+			continue
+		}
+		if event.Type == "DELETED" {
+			isCnsFileAccessConfigInstanceDeleted = true
+		}
+		cnsFileAccessConfigInstanceErr = cnsfileaccessconfig.Status.Error
```

Comment:
When the object is deleted, will the `cnsfileaccessconfig` instance received on line 618 be empty?

Reply:
No